### PR TITLE
SWIG: Add DotNET Native Interop support

### PR DESCRIPTION
**Pull request by Jorteck**
This PR implements the generation of bindings that are needed to support Managed <-> Native interop in DotNET. The Managed implementation can be found here: https://github.com/nwn-dotnet/NWN.Native/tree/main/src/main/API

I'd like some feedback on this and possibly some help migrating the "generate.sh" script to a SWIG CMake task: https://cmake.org/cmake/help/latest/module/UseSWIG.html

### CODE REVIEW

File: NWNXLib/API/API/CNWSDialog.hpp [hunk 1]
```diff
@@ -61,15 +62,15 @@ struct CNWSDialog
...
+    BOOL GetStartEntryOneLiner(CNWSObject * pNWSObjectOwner, CExoLocString & sOneLiner, CResRef & sSound, CResRef & sScript, CExoArrayList<ScriptParam>& scriptParams);
     float SetDialogDelay(CNWSObject * pNWSObjectOwner, CExoLocString sLocText, uint32_t nMinSeconds, BOOL bEnableNPCTiming);
     void AddJournalEntry(const CExoString & szPlotID, uint32_t nState, uint32_t nPlayerId);
     BOOL SendDialogEntry(CNWSObject * pNWSObjectOwner, uint32_t nPlayerIdGUIOnly, uint32_t iEntry, BOOL bPlayHelloSound);
     BOOL SendDialogReplies(CNWSObject * pNWSObjectOwner, uint32_t nPlayerIdGUIOnly);
     BOOL HandleReply(uint32_t nPlayerID, CNWSObject * pNWSObjectOwner, uint32_t nReplyIndex, BOOL bEscapeDialog, uint32_t currentEntryIndex);
     BOOL RemovePlayer(OBJECT_ID oidRemove);
     BOOL IsPlayerInDialog(OBJECT_ID oidPlayer);
-    void LoadScriptParams(CResGFF * pRes, ScriptParamsList & paramsList, CResStruct * pStructure, char * szFieldID);
+    void LoadScriptParams(CResGFF * pRes, CExoArrayList<ScriptParam>& paramsList, CResStruct * pStructure, char * szFieldID);
```
Comment (mtijanic) [inline NWNXLib/API/API/CNWSDialog.hpp:73]: +1, the old way didn't match the mangled name.

File: NWNXLib/API/API/CNWSScriptVar.hpp [hunk 2]
```diff
@@ -23,7 +23,6 @@ struct CNWSScriptVar
     CScriptLocation m_location;
 
     CNWSScriptVar();
-    CNWSScriptVar(const CNWSScriptVar &);
```
Comment (mtijanic) [inline NWNXLib/API/API/CNWSScriptVar.hpp:26]: Ugh, this whole class is missing from FunctionsLinux.hpp. The API _is_ correct. I'll add the offsets.

File: NWNXLib/API/API/CTlkTable.hpp [hunk 3]
```diff
@@ -47,7 +47,7 @@ struct CTlkTable
     virtual void GetTokenValue(const CTlkTableToken & cTlkTableToken, CExoString & sToken);
     BOOL FetchInternal(uint32_t strId, STR_RES & strRes, BOOL bParse);
     void CloseFile(CTlkFile * pTlkFile);
-    BOOL OpenFile(const CExoString & sFileName, CTlkFile * pTlkFile);
+    BOOL OpenFile(const CExoString & sFileName, CTlkFile ** pTlkFile);
```
Comment (mtijanic) [inline NWNXLib/API/API/CTlkTable.hpp:50]: Technically, it's `CTlkFile *pTlkFile[2]`, and the CloseFile() should be updated as well.

File: Plugins/DotNET/CMakeLists.txt [hunk 4]
```diff
@@ -1,3 +1,14 @@
...
-    "DotNET.cpp"
-    "DotNETExports.cpp")
+add_custom_target(
+    DotNETGenerate
+    COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/generate.sh
+    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+)
+
+add_library(DotNET SHARED "DotNET.cpp" "DotNETExports.cpp" "DotNETInterop.cpp")
+set_property(SOURCE ${CMAKE_CURRENT_SOURCE_DIR}/DotNETInterop.cpp PROPERTY GENERATED 1)
```
Comment (mtijanic) [inline Plugins/DotNET/CMakeLists.txt:8]: Can you post what the generated .cpp file looks like?

**Comment by mtijanic** [comment]
I merged the API fixup separately, so we can keep this more ontopic.

File: Plugins/DotNET/CMakeLists.txt [hunk 5]
```diff
@@ -1,3 +1,3 @@
 add_plugin(DotNET
     "DotNET.cpp"
-    "DotNETExports.cpp")
+    "DotNETExports.cpp")
```
Comment (mtijanic) [inline Plugins/DotNET/CMakeLists.txt:3]: Should probably revert all changes to DotNET/ - ere, and DotNET/DotNET.i

File: Plugins/SWIG/DotNET.i [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+%template(CResHelperIFO) CResHelper<CResIFO,2014>;
+%template(CResHelperSET) CResHelper<CResSET,2013>;
+%template(CResHelperWOK) CResHelper<CResWOK,2016>;
+%template(CResHelperNSS) CResHelper<CResNSS,2009>;
+%template(CResHelperTLK) CResHelper<CResTLK,2018>;
+%template(CResHelperNDB) CResHelper<CResNDB,2064>;
+%template(CResHelperNCS) CResHelper<CResNCS,2010>;
+
+// Ignore APIs with no symbols.
+%ignore CNWSObject::BindEventToHandler;
```
Comment (mtijanic) [inline Plugins/SWIG/DotNET.i:71]: All of these should no longer be necessary.

File: Plugins/SWIG/NWNXLib.i [hunk 7]
```diff
@@ -0,0 +1,671 @@
+%{
+#include "API/ALL_CLASSES.hpp"
+#include "API/AurList.hpp"
```
Comment (mtijanic) [inline Plugins/SWIG/NWNXLib.i:3]: You'll need to regenerate this list, I deleted quite a few of the headers..

**Review by mtijanic — COMMENTED** [review]
I'm not sure if we should generate the .cpp file every build, or simply keep it checked in along with generate.sh and run it manually. API updates are really rare, so it's no issue to do it manually, and saves everyone a dependency during build time.. and makes cmake much simpler. What do you think?

**Comment by Jorteck** [comment]
> I'm not sure if we should generate the .cpp file every build, or simply keep it checked in along with generate.sh and run it manually. API updates are really rare, so it's no issue to do it manually, and saves everyone a dependency during build time.. and makes cmake much simpler. What do you think?

I wouldn't mind keeping it for validation of the API. Is the extra build time significant?

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -24,7 +24,7 @@ env:
 
 jobs:
   build:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-20.04
```
Comment (mtijanic) [inline .github/workflows/build.yml:27]: Can you explain why this is needed? It sounds like a good idea to use a fixed version regardless, but it also sounds like it should be a separate PR altogether.

File: Plugins/DotNET/CMakeLists.txt (hunk 5 above)
Comment (mtijanic) [reply Plugins/DotNET/CMakeLists.txt:3]: I'm still seeing DotNET/DotNET.i file and a change (removed newline) in DotNET/CMakeLists.txt.

File: Plugins/SWIG/generate.sh [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+echo "%}" >> $NWNXLIB_SWG
+
+for include in "${INCLUDES[@]}"
+do
+    echo "%include \"$include\"" >> $NWNXLIB_SWG
+done
+
+find ../../NWNXLib/API -type f -name '*.hpp' "${FIND_ARGS[@]}" -printf '%%include "%P"\n' >> $NWNXLIB_SWG
+
+swig ${@:1}
```
Comment (mtijanic) [inline Plugins/SWIG/generate.sh:22]: Please configure your editor to produce a newline at the end of files. There are plenty of tools that get confused by this, and also most languages' standards explicitly require it (even if the actual implementation can deal with it missing)

File: Plugins/SWIG/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    set_property(SOURCE ${CMAKE_CURRENT_SOURCE_DIR}/${interopfile} PROPERTY GENERATED 1)
+    add_sanitizers(${target})
+    target_link_libraries(${target} Core "-Wl,--no-undefined")
+    set_target_properties(${target} PROPERTIES PREFIX "${PLUGIN_PREFIX}")
+    target_include_directories(${target} PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}")
+    target_compile_definitions(${target} PRIVATE "-DPLUGIN_NAME=\"${PLUGIN_PREFIX}${target}\"")
+    add_dependencies(${target} Generate_${target})
+endfunction()
+
+add_swig_plugin(SWIG_DotNET "DotNETInterop.cpp" -csharp -dllimport "${PLUGIN_PREFIX}SWIG_DotNET" -namespace "NWN.Native.API" "DotNET.i")
```
Comment (mtijanic) [inline Plugins/SWIG/CMakeLists.txt:18]: If we are going to keep generating it every build, rather than do manually, we need two more things:
1. Skip the plugin altogether if people don't have the swig dependency installed. Adding a new plugin should never break people's builds.
2. Set up make dependencies on the API, so that it is only rebuilds the generated files if the source API files changed. If I'm reading this right, it will generate DotNETInterop.cpp every time I edit an unrelated file in a different plugin and try to build it.

**Review by mtijanic — COMMENTED** [review]
I think this also needs the actual barebones plugin body that exposes `PluginLoad`, otherwise the core will throw an error when loading it. See Scripts/bootstrap_plugin.sh for the minimal(ish) body needed.

File: .github/workflows/build.yml (hunk 8 above)
Comment (Jorteck) [reply .github/workflows/build.yml:27]: ubuntu-latest (18.04) only has SWIG 3.0, and has trouble parsing static_asserts: https://github.com/nwnxee/unified/runs/1677788148#step:10:83.

20.04 fixes this as it has access to SWIG 4.
